File: Graph Algorithms/Building Teams (1668)/Dukkha_3.java
// https://cses.fi/problemset/task/1668
// Building Teams
// practice with rainboy
import java.io.*;
import java.util.*;

public class CSES1668 {
    static ArrayList[] aa;
    static int[] cc;
    static int[] stack;
    static boolean dfs(int i, int c) {
        int cnt = 0;
        cc[i] = c;
        stack[cnt++] = i;
        while (cnt > 0) {
            i = stack[--cnt];
            c = 3 - cc[i];
            ArrayList<Integer> adj = aa[i];
            for (int j : adj)
                if (cc[j] == 0) {
                    cc[j] = c;
                    stack[cnt++] = j;
                } else if (cc[j] != c)
                    return false;
        }
        return true;
    }
    public static void main(String[] args) throws IOException {
        _Scanner sc = new _Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        aa = new ArrayList[n];
        for (int i = 0; i < n; i++)
            aa[i] = new ArrayList<Integer>();
        while (m-- > 0) {
            int i = sc.nextInt() - 1;
            int j = sc.nextInt() - 1;
            aa[i].add(j);
            aa[j].add(i);
        }
        cc = new int[n];
        stack = new int[n];
        boolean yes = true;
        for (int i = 0; i < n; i++)
            if (cc[i] == 0 && !dfs(i, 1)) {
                yes = false;
                break;
            }
        PrintWriter pw = new PrintWriter(System.out);
        if (yes) {
            for (int i = 0; i < n; i++)
                pw.print(cc[i] + " ");
            pw.println();
        } else
            pw.println("IMPOSSIBLE");
        pw.close();
    }
    static class _Scanner {
        InputStream is;
        _Scanner(InputStream is) {
            this.is = is;
        }
        byte[] bb = new byte[1 << 15];
        int k, l;
        byte getc() throws IOException {
            if (k >= l) {
                k = 0;
                l = is.read(bb); //if (l < 0) return -1;
            }
            return bb[k++];
        }
        byte skip() throws IOException {
            byte b;
            while ((b = getc()) <= 32)
                ;
            return b;
        }
        String next() throws IOException {
            StringBuilder sb = new StringBuilder();
            for (byte b = skip(); b > 32; b = getc())
                sb.append((char) b);
            return sb.toString();
        }
        int nextInt() throws IOException {
            int n = 0;
            for (byte b = skip(); b > 32; b = getc())
                n = n * 10 + b - '0';
            return n;
        }
    }
}

File: Additional Problems/Creating Offices (1752)/Dukkha_3.cpp
// https://cses.fi/problemset/task/1752
// Creating Offices
// practice with rainboy
#include <iostream>
#include <vector>

using namespace std;

const int INF = 0x3f3f3f3f, N = 200000;

vector<int> aa[N];
bool used[N];
int cc[N], dd[N], dd_[N], lambda;

void dfs(int p, int i, int d) {
    dd[i] = d;
    int d_ = INF, c_ = -1;
    for (int j : aa[i])
        if (j != p) {
            dfs(i, j, d + 1);
            d_ = min(d_, dd_[j]);
            int c = cc[j];
            if (c == -1)
                continue;
            if ((dd[c] - d) * 2 >= lambda) {
                used[c] = true;
                d_ = min(d_, dd[c]);
            } else if (c_ == -1 || dd[c_] < dd[c])
                c_ = c;
        }
    dd_[i] = d_;
    if (c_ != -1 && dd[c_] - d + d_ - d < lambda)
        c_ = -1;
    if (c_ == -1 && d_ - d >= lambda)
        c_ = i;
    cc[i] = c_;
}

int main() {
    int n;
    scanf("%d%d", &n, &lambda);
    for (int h = 0; h < n - 1; h++) {
        int i, j;
        scanf("%d%d", &i, &j), i--, j--;
        aa[i].push_back(j);
        aa[j].push_back(i);
    }
    dfs(-1, 0, 0);
    if (cc[0] != -1)
        used[cc[0]] = true;
    int k = 0;
    for (int i = 0; i < n; i++)
        if (used[i])
            k++;
    printf("%d\n", k);
    for (int i = 0; i < n; i++)
        if (used[i])
            printf("%d ", i + 1);
    printf("\n");
    return 0;
}
